### PR TITLE
#303: fix clean_tags with tags dict

### LINKED ISSUE #303: Failing to clean up tag vocabularies
The configuration
```
  {"clean_tags": true}
```
will break the harvester in case remote tags are imported as dict and not plain strings.

The error is:
```
2017-09-19 10:08:43,652 ERROR  [ckanext.harvest.harvesters.base] [base.py:371] -- ord() expected a character, but string of length 13 found
Traceback (most recent call last):
  File "/usr/lib/ckan/default/src/ckanext-harvest/ckanext/harvest/harvesters/base.py", line 291, in _create_or_update_package
    tags = [munge_tag(t) for t in tags if munge_tag(t) != '']
  File "/usr/lib/ckan/default/src/ckan/ckan/lib/munge.py", line 108, in munge_tag
    tag = substitute_ascii_equivalents(tag)
  File "/usr/lib/ckan/default/src/ckan/ckan/lib/munge.py", line 98, in substitute_ascii_equivalents
    if ord(char) in char_mapping:
TypeError: ord() expected a character, but string of length 13 found
```

and it happens because the `munge` function is looping not on the [legacy string format](https://github.com/ckan/ckanext-harvest/blob/v1.1.0/ckanext/harvest/harvesters/base.py#L231-L235), but [on a dict](https://github.com/ckan/ckanext-harvest/blob/v1.1.0/ckanext/harvest/harvesters/base.py#L244-L245) like this:
```
{u'vocabulary_id': None, u'state': u'active', u'display_name': u'Zone a rischio naturale', u'id': u'073080c8-fef2-4743-9c9e-6216019f8b3d', u'name': u'Zone a rischio naturale'}
```
so the var `char` will be holding the `vocabulary_id` string.

(Initally I was thinking this issue was related to #154, but atm tags from vocabularies are properly imported.)

### PR DESCRIPTION
**Pull request by etj**
Fix #303

### CODE REVIEW

File: ckanext/harvest/harvesters/base.py [hunk 1]
```diff
@@ -288,9 +288,21 @@ def _create_or_update_package(self, package_dict, harvest_object,
 
             if self.config and self.config.get('clean_tags', False):
                 tags = package_dict.get('tags', [])
-                tags = [munge_tag(t) for t in tags if munge_tag(t) != '']
-                tags = list(set(tags))
-                package_dict['tags'] = tags
+                if tags:
+                    if not isinstance(tags[0], dict):
```
Comment (metaodi) [inline ckanext/harvest/harvesters/base.py]: Please do not check the type like this (and the `if tags:` is not necessary). Python uses the [EAFP style](http://python.net/~goodger/projects/pycon/2007/idiomatic/handout.html#eafp-vs-lbyl), so you should rather wrap this in a try/except and handle the exception, something like this:

```python
try:
    # assume it's in the package_show form
   tags = [munge_tag(t['name']) for t in tags if munge_tag(t['name']) != '']
   ...
except TypeError: # a TypeError is raised if `t` above is a string
   # REST format: 'tags' is a list of strings
   tags = [munge_tag(t) for t in tags if munge_tag(t) != '']
   ...

File: ckanext/harvest/harvesters/base.py [hunk 2]
```diff
@@ -288,9 +288,21 @@ def _create_or_update_package(self, package_dict, harvest_object,
...
+                        # REST format: 'tags' is a list of strings
+                        tags = [munge_tag(t) for t in tags if munge_tag(t) != '']
+                        tags = list(set(tags))
+                        package_dict['tags'] = tags
+                    else:
+                        # package_show form: 'tags' is a list of dicts
+                        clean_tags = []
+                        for t in tags:
```
Comment (metaodi) [inline ckanext/harvest/harvesters/base.py]: Use a list comprehension just like above, no need for a `for` loop here.

Comment (etj) [reply ckanext/harvest/harvesters/base.py]: Uhm ok, but we're updating a field in a dict, guess I'll need something more complex

File: ckanext/harvest/harvesters/base.py (hunk 1 above)
Comment (etj) [reply ckanext/harvest/harvesters/base.py]: ok

**Comment by etj** [comment]
WIll also move the logic for cleaning the tags in its own method for better testing it

**Comment by metaodi** [comment]
Thank you @etj

**Pull request merged**